lib/tests/test_rewrite_transform.rs

```diff
@@ -120,3 +122,89 @@ fn test_transform_descendants_sync_linearize_merge() {
 
     assert_eq!(new_commit_c.parent_ids(), vec![commit_b.id().clone()]);
 }
+
+// Reorder commits B and C by using the `parents_callback`. Reordering has to be
+// done outside of the typical callback since we must ensure that the new
+// traversal order of the commits is valid.
+//
+// G
+// | E
+// | D F
+// | |/
+// | C
+// | B
+// |/
+// A
+#[test]
+fn test_transform_descendants_parent_callback() {
+    let test_repo = TestRepo::init();
+    let repo = &test_repo.repo;
+
+    let mut tx = repo.start_transaction();
+    let mut graph_builder = CommitGraphBuilder::new(tx.repo_mut());
+    let commit_a = graph_builder.initial_commit();
+    let commit_b = graph_builder.commit_with_parents(&[&commit_a]);
+    let commit_c = graph_builder.commit_with_parents(&[&commit_b]);
+    let commit_d = graph_builder.commit_with_parents(&[&commit_c]);
+    let commit_e = graph_builder.commit_with_parents(&[&commit_d]);
+    let commit_f = graph_builder.commit_with_parents(&[&commit_c]);
+    let commit_g = graph_builder.commit_with_parents(&[&commit_a]);
+
+    let options = RewriteRefsOptions::default();
+    let mut rebased = HashMap::new();
+    tx.repo_mut()
+        .transform_descendants_with_options(
+            vec![commit_b.id().clone()],
+            &options,
+            |commit, _| {
+                if *commit == commit_c {
+                    Some(vec![commit_a.id().clone()])
+                } else if *commit == commit_b {
+                    Some(vec![commit_c.id().clone()])
+                } else {
+                    Some(
+                        commit
+                            .parent_ids()
+                            .iter()
+                            .map(|parent_id| {
+                                if parent_id == commit_c.id() {
+                                    commit_b.id().clone()
+                                } else {
+                                    parent_id.clone()
+                                }
+                            })
+                            .collect_vec(),
+                    )
+                }
+            },
```

Comment:
Random thought: Does it make sense to pass `HashMap<CommitId, Vec<CommitId>>` to just reorder the `to_visit` stack? In this example, `{C: [A], B: [C]}` would be passed as parent overrides, and the last `else` block would be handled by the rewrite callback.

My feeling is that the callback API is harder to do things right, but I have no idea if overrides map will help simplify API surface.

Reply:
Yes, I was considering that too, but didn't try it out yet. I'll try to create a version which uses a map instead.

I do agree that the callback is quite messy, and might be error-prone.

Reply:
I've updated this PR to accept a `new_parents_map`, which might make it more convenient to construct the lsit of new parents.